src/compiler/debug.ts

```diff
@@ -410,5 +470,92 @@ namespace ts {
+                }
+            };
+        }
+
+        function createDeprecation(name: string, options: DeprecationOptions & { error: true }): () => never;
+        function createDeprecation(name: string, options?: DeprecationOptions): () => void;
+        function createDeprecation(name: string, options: DeprecationOptions = {}) {
+            const version = typeof options.typeScriptVersion === "string" ? new Version(options.typeScriptVersion) : options.typeScriptVersion ?? getTypeScriptVersion();
+            const errorAfter = typeof options.errorAfter === "string" ? new Version(options.errorAfter) : options.errorAfter;
+            const warnAfter = typeof options.warnAfter === "string" ? new Version(options.warnAfter) : options.warnAfter;
+            const since = typeof options.since === "string" ? new Version(options.since) : options.since ?? warnAfter;
+            const error = options.error || errorAfter && version.compareTo(errorAfter) <= 0;
+            const warn = !warnAfter || version.compareTo(warnAfter) >= 0;
+            return error ? createErrorDeprecation(name, errorAfter, since, options.message) :
+                warn ? createWarningDeprecation(name, errorAfter, since, options.message) :
+                noop;
+        }
+
+        function wrapFunction<F extends (...args: any[]) => any>(deprecation: () => void, func: F): F {
+            return function (this: unknown) {
+                deprecation();
+                return func.apply(this, arguments);
+            } as F;
+        }
+
+        function wrapAccessor(deprecation: () => void, desc: PropertyDescriptor) {
+            if (desc.configurable) {
+                const newDesc: PropertyDescriptor = { ...desc, enumerable: false };
+                if (desc.get) newDesc.get = wrapFunction(deprecation, desc.get);
+                if (desc.set) newDesc.set = wrapFunction(deprecation, desc.set);
+                return newDesc;
+            }
+        }
+
+        function wrapValue(deprecation: () => void, desc: PropertyDescriptor) {
+            if (typeof desc.value === "function" && (desc.configurable || desc.writable)) {
+                const newDesc: PropertyDescriptor = { ...desc };
+                if (desc.configurable) {
+                    desc.enumerable = false;
+                }
+                newDesc.value = wrapFunction(deprecation, newDesc.value);
+                return newDesc;
+            }
+        }
+
+        export function deprecateExport<T, K extends Extract<MatchingKeys<T, (...args: any[]) => any>, string>>(ns: T, key: K, options?: DeprecationOptions) {
+            const desc = Object.getOwnPropertyDescriptor(ns, key);
+            if (!desc) return;
+            const deprecation = createDeprecation(key, options);
+            const newDesc = desc.get || desc.set ? wrapAccessor(deprecation, desc) : wrapValue(deprecation, desc);
+            if (newDesc) {
+                Object.defineProperty(ns, key, newDesc);

```

Comment:
This is probably going to have a bad time when refactored to modules; I think the `defineProperty` will throw when the export already exists as a non-configurable getter (ie, downlevel), and will likewise throw in native modules. It might be a good idea to preemptively change it to a usage pattern like
```ts
export const updateGetAccessor = Debug.deprecate(factory.updateGetAccessor, { since: "4.0", warnOn: "4.1", message });
```
to avoid the need for reflective reassignment. If you pull off the `message` (or the entire options object) into a local variable, it's not much more verbose than the current export+deprecate setup, but keeps the exports and deprecation for the exports on the same line (so we're not going to accidentally remove one but not the other later).